### ISSUE TITLE
Cannot start repair on specific column family if a schedule exists

### ISSUE THREAD

**Issue by ecnepsnai** [thelastpickle/cassandra-reaper#971]
We have a schedule set up for a repair on all CFs on a keyspace, but we needed to repair just a single table. I paused the schedule and when the last repair from it finished I tried to submit a manual repair run just for a single table, but I get the following exception:

```
java.lang.IllegalArgumentException: unit conflicts with existing in <cluster>:<keyspace>
```

There was a similar issue reported with https://github.com/thelastpickle/cassandra-reaper/issues/849

However this is slightly different - The repair I am trying to start is not the same as the schedule, and the schedule itself is paused. I had to delete the schedule entirely to get it to work.

┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-461) by [Unito](https://www.unito.io)
┆Issue Number: K8SSAND-461

**Comment by adejanovski** [comment]
Hi, 

Reaper indeed disallows creating repairs which will [overlap on at least one table](https://github.com/thelastpickle/cassandra-reaper/blob/bc9c738c5a31db1a515cc0f2f7683c8b955234a4/src/server/src/main/java/io/cassandrareaper/service/RepairUnitService.java#L169-L170).
@michaelsembwever, I think we should be more flexible on this as there are cases where you'll want to repair a specific table for various reasons, and not the whole cluster.
If we don't want to allow this by default, maybe we could add a setting to override the behavior?
Same for allowing incremental and full repairs on the same keyspace. For this, we could block it for versions of Cassandra where this could be harmful and allow it for recent versions (3.11.4 and >, or something like this. I'd need to check the Cassandra Jira).

**Comment by michaelsembwever** [comment]
> Reaper indeed disallows creating repairs which will overlap on at least one table.

Yes, we are being opinionated here preventing users from blowing off their feet from repairing tables in multiple ways.

Before Cassandra 4.0 (some might say the latest 3.11.x but i disagree, as there's not enough feedback and evidence on the table yet) full and incremental repairs should not be mixed.

There are also repair options that operators may want to enforce: segment size, intensity, dc_aware vs sequential; ensuring other Reaper users don't come along and "do it a different way". There's also the issue of tables that must not be repaired, like through using MVs or CDC.

This does need to evolve, and needs a bit more discussion around it. It would be good to also reference the past discussions we have had on this, so we're all on the same page and not repeating ourselves. (I had a quick search but couldn't find the related issues/PRs.)

For the meantime I would recommend repairing the column family with the existing schedule you have. Repairing additional tables shouldn't be a problem. (If it is you need to address that ;) 
Otherwise break that existing schedule up.

**Comment by eriksw** [comment]
It's not right at all for this to include schedules that are disabled.

Scenario: I have a schedule whose normal run has been stuck for a week. I want to do a manual 1000-segment-per-node repair on that table to force reaper to take care of the issue. I am prevented from doing that without deleting the schedule.

**Comment by michaelsembwever** [comment]
> It's not right at all for this to include schedules that are disabled.

There is not currently any way to disable a schedule.

> Scenario: I have a schedule whose normal run has been stuck for a week. I want to do a manual 1000-segment-per-node repair on that table to force reaper to take care of the issue. I am prevented from doing that without deleting the schedule.


A suggestion to make here is that the schedule is in fact configured wrong and it could and should be better defined.

For example, a workaround to solve this:
1. Take that table out of the schedule.
2. Manually repair the table. Do it a couple of times to tune the correct segment-per-node value.
3. Create a new schedule for that table and its new segment-per-node value.

This should put you in a more robust position.

> This does need to evolve, and needs a bit more discussion around it.

This still stands. If you have more heuristics to share they can only help.

**Comment by eriksw** [comment]
> There is not currently any way to disable a schedule.

👎 There is most certainly a "Stop" / "Activate" toggle in the UI that definitely stops the schedule from starting any runs.

> 1. Take that table out of the schedule.

👎 There's no UI for editing existing schedules without deleting then re-creating them.

When the desired manual run is conflicting with a schedule that has a nontrivial set of carefully hand-picked tables, having to delete it then re-select them is extremely burdensome.

> 2. Manually repair the table. Do it a couple of times to tune the correct segment-per-node value.

👎 That is **what I am trying to do** by creating the repair that this restriction prohibits.

**Comment by michaelsembwever** [comment]
> There is most certainly a "Stop" / "Activate" toggle in the UI that definitely stops the schedule from starting any runs.

I stand corrected! It's only been there for 5 years 🤣

Being able to run repairs that overlap on a disabled schedule could be one way forward. Another possibility is introducing a 'force' option. Again it's part of the bigger discussion to be had.

I'm still leaning towards the idea of replacing the restriction with a pop-up warning dialog… something that informs the user a schedule for these repairs is already in place and it repairs using different options, and does the user really want to continue. 


 > There's no UI for editing existing schedules without deleting then re-creating them.

That's a good idea! You are free to implement and provide a patch for that. That would be most welcome. 

I meant deleting the schedule, and re-creating it. It may even be that separating the troublesome table is advantageous for you in the long run.

>That is what I am trying to do by creating the repair that this restriction prohibits. 

Hopefully the workaround has been better explained now.

**Comment by eriksw** [comment]
If it was just a warning that can be bypassed, instead of a hard restriction, I think it would be entirely reasonable to show it even when the conflict is with a disabled schedule.

**Comment by adejanovski** [comment]
Hi @eriksw,

I agree with @michaelsembwever that it's good to enforce best practices by default, but I hear your frustration and reckon we should allow experienced users to bypass such limitations.

Also, we really need to give the ability to update existing schedules, it's a big UX issue (created #1006 to track this) 

@michaelsembwever, what do you think about adding config settings to disable some overlap checks between repair units? By default we keep the checks as they currently are (although we'll revisit some of them with 4.0) while giving more flexibility for one shot cases such as @eriksw's.

**Comment by michaelsembwever** [comment]
> @michaelsembwever, what do you think about adding config settings to disable some overlap checks between repair units? By default we keep the checks as they currently are (although we'll revisit some of them with 4.0) while giving more flexibility for one shot cases such as @eriksw's.

I would rather the pop-up warning. Even experienced users will still benefit from the feedback, and it could well prevent some disasters (even for experienced users). With a nice UI that reaper has it just becomes too easy to click away and not double-check things carefully enough. Is a pop-up dialog difficult to do? (Must be easier than the editing schedule requirement, which is another obvious pain point.)

**Comment by adejanovski** [comment]
I guess we could do both.
Currently we're imposing to do one off repairs outside of Reaper, which can be very challenging and lead to more problems due to anticompaction.
I'd rather have those run in Reaper as well without having to destroy existing schedules.

**Comment by eriksw** [comment]
If it is made a config option (please don't go this route), please be sure it's configurable by Docker users via _only_ an environment variable with no file changes needed. Config that requires injecting a file effectively does not exist.

**Comment by michaelsembwever** [comment]
> I guess we could do both.

We could, but I rather we didn't.

**Comment by adejanovski** [comment]
Re-read the whole thread and understood your position a little better I think @michaelsembwever. 

The follow up task here would be:

- Make the schedules editable in the UI (instead of having to delete and recreate): #1006
- Allow overlapping repair units but issue a warning in the UI, notifying that another schedule/repair overlaps. Provide buttons to either force or cancel the repair run. This warning shouldn't pop up if the overlapping schedule is about the type of repair (full/incremental) starting with 4.0.

Sounds good?

**Comment by michaelsembwever** [comment]
> This warning shouldn't pop up if the overlapping schedule is about the type of repair (full/incremental) starting with 4.0.

Let's discuss that. I think if the dialog is worded correctly is should always appear.
The feedback of the comparison to how this (these) tables are normally repaired versus what you are about to do, is always useful. But pre 4.0 it's definitely more a warning and going against our recommendations, than just confirmation.

**Comment by adejanovski** [comment]
@michaelsembwever, @eriksw, let me know what you think of #1008.

**Comment by eriksw** [comment]
@adejanovski Looks like #1008 was closed for #1078, but #1078 did not address the need to force a manual repair that overlaps with a schedule.

We still need the ability to do that for all the same reasons as previously discussed in this issue. Namely, to un-stick schedules that already have high-four-figure segment counts but with segments that just time out over and over and over.

**Comment by adejanovski** [comment]
@eriksw, I've noticed this while working on another PR and made the change there to auto-force manual repairs: https://github.com/thelastpickle/cassandra-reaper/pull/1107/files#diff-137a33700d214bb76d0a57b8da0b5aa2cc84fb998abedbe6cde7822da1df3e26R216-R217

It'll be merged soon hopefully.

**Comment by adejanovski** [comment]
These features are all available in 3.1.
